hugr-passes/src/dataflow/test.rs

```diff
@@ -0,0 +1,310 @@
+
+    let mut machine = Machine::default();
+    machine.propolutate_out_wires([(r_w, PartialValue::new_variant(3, []))]);
+    machine.run(TestContext(Arc::new(&hugr)));
+
+    let o_r = machine
+        .read_out_wire(tl_o)
+        .unwrap()
+        .try_into_wire_value(&hugr, tl_o)
+        .unwrap();
+    assert_eq!(o_r, r_v);
+    assert_eq!(
+        Some(TailLoopTermination::NeverContinues),
+        machine.tail_loop_terminates(&hugr, tail_loop.node())
+    )
+}
+
+#[test]
+fn test_tail_loop_always_iterates() {
+    let mut builder = DFGBuilder::new(Signature::new_endo(vec![])).unwrap();
+    let r_w = builder
+        .add_load_value(Value::sum(0, [], SumType::new([type_row![], BOOL_T.into()])).unwrap());
+    let true_w = builder.add_load_value(Value::true_val());
+
+    let tlb = builder
+        .tail_loop_builder([], [(BOOL_T, true_w)], vec![BOOL_T].into())
+        .unwrap();
+
+    // r_w has tag 0, so we always continue;
+    // we put true in our "other_output", but we should not propagate this to
+    // output because r_w never supports 1.
+    let tail_loop = tlb.finish_with_outputs(r_w, [true_w]).unwrap();
+
+    let [tl_o1, tl_o2] = tail_loop.outputs_arr();
+    let hugr = builder.finish_hugr(&EMPTY_REG).unwrap();
+
+    let mut machine = Machine::default();
+    machine.run(TestContext(Arc::new(&hugr)));
+
+    let o_r1 = machine.read_out_wire(tl_o1).unwrap();
+    assert_eq!(o_r1, PartialValue::bottom());
+    let o_r2 = machine.read_out_wire(tl_o2).unwrap();
+    assert_eq!(o_r2, PartialValue::bottom());
+    assert_eq!(
+        Some(TailLoopTermination::NeverBreaks),
+        machine.tail_loop_terminates(&hugr, tail_loop.node())
+    );
+    assert_eq!(machine.tail_loop_terminates(&hugr, hugr.root()), None);
+}
+
+#[test]
+fn test_tail_loop_iterates_twice() {
+    let mut builder = DFGBuilder::new(Signature::new_endo(vec![])).unwrap();
+    // let var_type = Type::new_sum([type_row![BOOL_T,BOOL_T], type_row![BOOL_T,BOOL_T]]);
+
+    let true_w = builder.add_load_value(Value::true_val());
+    let false_w = builder.add_load_value(Value::false_val());
+
+    // let r_w = builder
+    //     .add_load_value(Value::sum(0, [], SumType::new([type_row![], BOOL_T.into()])).unwrap());
+    let tlb = builder
+        .tail_loop_builder_exts(
+            [],
+            [(BOOL_T, false_w), (BOOL_T, true_w)],
+            vec![].into(),
+            ExtensionSet::new(),
+        )
+        .unwrap();
+    assert_eq!(
+        tlb.loop_signature().unwrap().signature(),
+        Signature::new_endo(type_row![BOOL_T, BOOL_T])
+    );
+    let [in_w1, in_w2] = tlb.input_wires_arr();
+    let tail_loop = tlb.finish_with_outputs(in_w1, [in_w2, in_w1]).unwrap();
+
+    let hugr = builder.finish_hugr(&EMPTY_REG).unwrap();
+    // TODO once we can do conditionals put these wires inside `just_outputs` and
+    // we should be able to propagate their values...ALAN wtf? loop control type IS bool ATM
```

Comment:
@doug-q please excuse the "wtf" but can you explain what you were planning here?? (They are currently in `inputs_outputs`, do you mean put them in both `just_inputs` and `just_outputs` ?)

Reply:
Well, I kept thinking that I could see what you were aiming for here, but it kept being harder than I hoped - the inputs to the second iteration are joined with the inputs to the first, so that tends to get you something uninformative.

But I kinda got there in the end - the loop input needs to be a *sum* with >0 (Partial)values in each variant so you can put different data in each. (And variant0 leads to doing another iteration with variant1, whereas variant1 leads to an exit.) Have added this as a separate test, it is quite cool that it does work.